lib/SILOptimizer/Utils/GenericCloner.cpp

```diff
@@ -106,24 +107,36 @@ void GenericCloner::populateCloned() {
         // Handle arguments for formal parameters.
         unsigned paramIdx = ArgIdx - origConv.getSILArgIndexOfFirstParam();
         if (ReInfo.isParamConverted(paramIdx)) {
-          // Store the new direct parameter to the alloc_stack.
-          createAllocStack();
+          assert(mappedType.isAddress());
+          mappedType = mappedType.getObjectType();
           auto *NewArg = ClonedEntryBB->createFunctionArgument(
               mappedType, OrigArg->getDecl());
-          getBuilder().createStore(Loc, NewArg, ASI,
-                                   StoreOwnershipQualifier::Unqualified);
 
-          // Try to create a new debug_value from an existing debug_value_addr.
+          // Try to create a new debug_value from an existing debug_value_addr
+          // for the argument. We do this before storing to ensure that when we
+          // are cloning code in ossa the argument has not been consumed by the
+          // store below.
           for (Operand *ArgUse : OrigArg->getUses()) {
             if (auto *DVAI = dyn_cast<DebugValueAddrInst>(ArgUse->getUser())) {
+              auto *oldScope = getBuilder().getCurrentDebugScope();
               getBuilder().setCurrentDebugScope(
                   remapScope(DVAI->getDebugScope()));
               getBuilder().createDebugValue(DVAI->getLoc(), NewArg,
                                             *DVAI->getVarInfo());
-              getBuilder().setCurrentDebugScope(nullptr);
+              getBuilder().setCurrentDebugScope(oldScope);
```

Comment:
@adrian-prantl @dcci @vedantk did I do this right? I don't understand how setting this to nullptr made sense before.

Reply:
It looks like GenericCloner::postProcess fixes up the scopes after populateCloned is finished. SILCloner isn't equipped to handle debug insts without scopes, hence the special-casing here. I'd expect you to be able to delete the second setCurrentDebugScope call.

I'm not sure why the cloning isn't structured in a way that allows remapping scopes as instructions get copied over.